### ISSUE TITLE
Allow values to be fetched lazily

### ISSUE THREAD

**Issue by helgoboss** [rossmacarthur/upon#10]
Hi, this crate looks great with its minimalistic yet scalable design, I would love to use it for [this](https://github.com/helgoboss/realearn/issues/683). I have a bit of a special case though: I can't collect all possible (leaf) values beforehand because I don't know which values are available, and it would also be suboptimal regarding performance. Fetching a single value is quick but fetching all is potentially expensive in my context.

That's why I would love to be able to pass a custom function that will be called by your rendering code and receives the fully qualified value key as parameter (e.g. `user.name`). The function then gets a chance to fetch the value and returns it.

I can also think of an alternative: Having a function in `Template` that returns the list of all keys used in the template. Then I could fetch exactly those values beforehand, put them into a map and provide them as value, no changes to the rendering process necessary. I think I would actually prefer this as it prevents repeated fetching of the same value and I need that list of keys anyway for other purposes. Plus, it's probably easier to implement, right?

What do you think of these two approaches? Would you consider adding them? You could also point me a bit into the right direction and I could try it myself. Thank you!

**Comment by rossmacarthur** [comment]
This is an interesting use case 😅. I am open to new features `upon`.

I do have some concerns how this would work with nested templates `{% include "nested" %}` 🤔? But this question aside, the following is how I might imagine an API for the two approaches:

### Custom value function

I think this method is more complete and easier for a user to understand the API but would require some refactoring of the renderer to be able to handle this. I think this could be accomplished by falling back to the value function in the case where no variable is found in the current scope.

```rust
impl Template {
    pub fn render_with_value_fn<F>(&self, ctx: F) -> Result<String>
    where
       F: Fn(&[&str]) -> Result<Value>,
}
```

One issue I can see is that it is possible the value function will need to understand which template is being rendered (in the case of nested templates)? So possibly we need to pass the template name

### Fetching keys

Fetching the variables that are referred to in a template is really easy but it has no context, so these paths could be referring to variables in the runtime scope only. The following would return all the referred to paths in a template.

```rust
impl Template {
    pub fn vars(&self) -> Vec<Vec<&str>> {
        self.template
            .instrs
            .iter()
            .filter_map(|instr| match instr {
                program::Instr::ExprStart(path) => Some(
                    path.iter()
                        .map(|ident| &self.source()[ident.span])
                        .collect(),
                ),
                _ => None,
            })
            .collect()
    }
}
```
E.g. given a template like this
```
{% for user in users %}
    {{ user.name }} - {{ user.age }}
{% endfor %}
```

Would return
```
[
    ["users"],
    ["user", "name"],
    ["user", "age"],
]
```

**Comment by helgoboss** [comment]
Wow, thanks for the detailed answer! I appreciate that very much.



> I do have some concerns how this would work with nested templates `{% include "nested" %}` 🤔?

I didn't even think of nested templates. It's a feature that I don't need in my case. Passing the template name as additional parameter to the function sounds reasonable to me though.

> Fetching the variables that are referred to in a template is really easy but it has no context

Oh, yet another thing I didn't think of ... so in your loop example, `user` is a kind of variable introduced by the template itself. Which means knowing that the template uses `["user", "name"]` doesn't really tell me anything because what `user` means depends on the context. I see.

This can only happen with loops, correct? In this case I would probably also simply refrain from using loops, they are not important for my use case as well :) Best thing if I could disable loops and nested templates, but I think then I would need to disable the complete block syntax which also disabled conditionals, mmh, but that's another question.

How would this loop context issue be dealt with in the _custom value function_ approach? If the function just receives `&[&str]`, there's no way it knows which item in the user vec is the current one. Maybe it can be something like this?

```rust
impl Template {
    pub fn render_with_value_fn<'a, F>(&self, ctx: F) -> Result<String>
    where
       F: Fn(&[Key<'a>]) -> Result<Value>,
}

pub enum Key<'a> {
    Normal(&'a str),
    Indexed(&'a str, usize),
}
```

It would for example pass:

1. `Key::Normal("users")`
2. `Key::Indexed("users", 0)`
3. `Key::Indexed("users", 1)`

Hope that makes sense.

**Comment by rossmacarthur** [comment]
> This can only happen with loops, correct?

Also the following
- [`with` blocks](https://github.com/rossmacarthur/upon/blob/trunk/SYNTAX.md#with)
- Includes blocks with an expression, e.g. `{% include "nested" with path.to.expr %}` statement, the `path.to.expr` would form the global context of the nested template.

> Best thing if I could disable loops and nested templates

Currently it is possible to disable block syntax (at *runtime*) by specifying a custom syntax that excludes block syntax. But like you said it disables conditionals as well.

```rust
let syntax = upon::SyntaxBuilder::new().expr("{{", "}}").build();
let engine = upon::Engine::with_syntax(syntax);
```

It might be possible to put each of these (includes, conditionals, loops) behind a feature flag so that they could be disabled at *compile time*.

> Key::Indexed("users", 1)

This would work for lists but I'm not sure how this would work for looping over maps 🤔, e.g. say you have a map of identifier to user in a `users` map. 

```
{% for id, user in users %}
   ...
{% endfor %}
```

> How would this loop context issue be dealt with in the custom value function approach? If the function just receives &[&str], there's no way it knows which item in the user vec is the current one.

🤔 I was imagining that the entire list would be returned as the expression to the `for` statement. E.g.

```
{% for user in users %}
               ^^^^^  <- the value fn called for this only 
               
     {{ user.name }} 
        ^^^^  <- renderer prioritizes local variables over calling the value fn so this just works
     
{% endfor %}
```

So the entire `users` list would be returned from the value fn (i.e. a `Value::List`). Then the renderer can just loop over it. This would easily work when `users` is a list or a map.

**Comment by helgoboss** [comment]
You are right, your approach how to deal with lists/maps makes much more sense. The "users" value should contain everything already, no need to fetch anything in there on demand. So there's no need to pass anything like "Key::Indexed". Definitely would be interested in that string-slice-based custom value function in the long term.

For the moment, disabling block syntax completely and having the "vars()" function in the Template struct would be a very big step forward for my use case already. So if you could add the "vars()" function, it would be awesome :D

**Comment by rossmacarthur** [comment]
So actually had some time to look into this and it looks like passing a custom function is maybe actually easier than I thought! Would you be willing to try it out and see if it fits your use case?

Use the following branch
```toml
# Cargo.toml
upon = { git = "https://github.com/rossmacarthur/upon.git", branch = "value-fn" }
```

Example usage:

```rust
use upon::ValueKey;

template.render_with_value_fn(|path| match path {
    &[ValueKey::Map("user"), ValueKey::Map("name")] => Ok(upon::Value::from("John Smith")),
     _ => Err(String::from("custom error message")),
})?;
```

**Comment by helgoboss** [comment]
Cool, thank you so much! I will have a look during the next days.

**Comment by rossmacarthur** [comment]
Fixed in #13 ([fb2a904](https://github.com/rossmacarthur/upon/commit/fb2a90444bd6da9baa15c9e8e9f378a9231e1d5f))
